File: src/test/java/ar/com/flow/bankaccount/BankAccountAssert.java
package ar.com.flow.bankaccount;

import ar.com.flow.bankaccount.balance.Balance;
import ar.com.flow.money.Money;
import org.assertj.core.api.AbstractAssert;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class BankAccountAssert extends AbstractAssert<BankAccountAssert, BankAccount> {
    public static BankAccountAssert assertThat(BankAccount account) {
        return new BankAccountAssert(account);
    }

    public BankAccountAssert(BankAccount account) {
        super(account, BankAccountAssert.class);
    }

    public BankAccountAssert keepsInitialBalance() {
        if (!actual.getBalance().equals(actual.getInitialBalance())) {
            failWithActualExpectedAndMessage(actual.getBalance(), actual.getInitialBalance(), "Balance should be equal to initial balance.");
        }

        return this;
    }

    public BankAccountAssert hasPositiveBalance(Money amount) {
        assertEquals(Balance.positive(amount), actual.getBalance());

        return this;
    }

    public BankAccountAssert increasedFunds(Money amount) {
        assertEquals(actual.getPreviousBalance().plus(amount), actual.getBalance());

        return this;
    }

    public BankAccountAssert decreasedFunds(Money amount) {
        assertEquals(actual.getPreviousBalance().minus(amount), actual.getBalance());

        return this;
    }

    public BankAccountAssert hasNegativeBalance(Money amount) {
        assertEquals(Balance.negative(amount), actual.getBalance());

        return this;
    }
}

File: src/test/kotlin/ar/com/flow/bankaccount/domain/BankAccountAssert.kt
package ar.com.flow.bankaccount.domain

import ar.com.flow.bankaccount.domain.balance.Balance.Companion.negative
import ar.com.flow.bankaccount.domain.balance.Balance.Companion.positive
import ar.com.flow.money.Money
import org.assertj.core.api.AbstractAssert
import org.junit.jupiter.api.Assertions.assertEquals

class BankAccountAssert(account: BankAccount) :
    AbstractAssert<BankAccountAssert, BankAccount>(account, BankAccountAssert::class.java) {
    fun keepsInitialBalance(): BankAccountAssert {
        if (actual.balance != actual.initialBalance) {
            failWithActualExpectedAndMessage(
                actual.balance,
                actual.initialBalance,
                "Balance should be equal to initial balance."
            )
        }
        return this
    }

    fun hasPositiveBalance(amount: Money): BankAccountAssert {
        assertEquals(positive(amount), actual.balance)
        return this
    }

    fun increasedFunds(amount: Money): BankAccountAssert {
        assertEquals(actual.previousBalance.plus(amount), actual.balance)
        return this
    }

    fun decreasedFunds(amount: Money): BankAccountAssert {
        assertEquals(actual.previousBalance.minus(amount), actual.balance)
        return this
    }

    fun hasNegativeBalance(amount: Money): BankAccountAssert {
        assertEquals(negative(amount), actual.balance)
        return this
    }

    companion object {
        fun assertThat(account: BankAccount): BankAccountAssert {
            return BankAccountAssert(account)
        }
    }
}